src/shogun/lib/any.h

```diff
@@ -318,8 +340,10 @@ namespace shogun {
 			enter_matrix(std::addressof(rows), std::addressof(cols));
 			if ((rows != _matrix->num_rows) || (cols != _matrix->num_cols))
 				*_matrix = SGMatrix<T>(rows, cols);
-			for (auto& _value: *_matrix)
-				on(std::addressof(_value));
+			for (auto index=0; index < cols; index++) {
+				on_matrix_row(std::addressof(rows), std::addressof(index), _matrix);
```

Comment:
I've changed a bit how the visitor behaves when dealing with SGMatrix (this way is much more customizable). @karlnapf @vigsterkr